services/keybase/keybase-btc.service.js

```diff
@@ -0,0 +1,98 @@
+'use strict'
+
+const KeybaseProfile = require('./keybase-profile')
+const Joi = require('joi')
+
+const bitcoinAddressFoundSchema = Joi.object({
+  them: Joi.array()
+    .items(
+      Joi.object({
+        cryptocurrency_addresses: {
+          bitcoin: Joi.array()
+            .items(
+              Joi.object({
+                address: Joi.string().required(),
+              }).required()
+            )
+            .required(),
+        },
+      }).required()
+    )
+    .min(0)
+    .max(1),
+}).required()
+
+const profileNotFoundSchema = Joi.object({
+  them: Joi.array().empty(),
+}).required()
+
+const bitcoinAddressFoundOrNotFound = Joi.alternatives(
+  bitcoinAddressFoundSchema,
+  profileNotFoundSchema
+)
+
+module.exports = class KeybaseBTC extends KeybaseProfile {
+  static get apiVersion() {
+    return '1.0'
+  }
+
+  static get category() {
+    return 'social'
+  }
+
+  static get route() {
+    return {
+      base: 'keybase/btc',
+      pattern: ':username',
+    }
+  }
+
+  static get defaultBadgeData() {
+    return { label: 'btc' }
+  }
+
+  async handle({ username }) {
+    const options = {
+      method: 'GET',
+      form: {
+        usernames: username,
+        fields: 'cryptocurrency_addresses',
+      },
+    }
+
+    const data = await this.fetch({
+      schema: bitcoinAddressFoundOrNotFound,
+      options,
+    })
+
+    try {
+      const address = data.them[0].cryptocurrency_addresses.bitcoin[0].address
+      return this.constructor.render({ address })
+    } catch (err) {
+      return {
+        message: 'not found',
```

Comment:
What sort of error responses could we see from the API? For example if a user specifies an invalid value for the `username` param, will the API return a 404 error? The default error message that will be displayed on 404 responses is `not found` so I'm thinking we may be able to provide more detailed/context specific error messages (for example something like `bitcoin address not found` here vs. `username not found` if the API returns a 404)

Reply:
The API returns 200 even if the profile doesn't exist.

See it by yourself:

* A valid profile (mine): https://keybase.io/_/api/1.0/user/lookup.json?usernames=skyplabs
* A non-existing profile: https://keybase.io/_/api/1.0/user/lookup.json?usernames=skyplabssssss

Reply:
Gotcha thanks. In that case, what if we did something like:

```
if (data.them[0] === null) {
  return {
    message: 'profile not found',
    color: 'critical'
  }
}

const bitcoinAddresses = data.them[0].cryptocurrency_addresses.bitcoin
if (bitcoinAddresses.length === 0) {
  return {
    message: 'no bitcoin addresses',
    color: 'inactive',
  }
}

return this.constructor.render({ address: bitcoinAddresses[0].address })

Reply:
I like this idea. However, I encountered some unexpected side effects while implementing it.

First, when I use the dashboard (the one on the screenshots) to generate a bitcoin badge, a request is sent for each letter of my username I type. My username being `skyplabs`, a first request is sent for `s`. The keybase API seems to return an HTTP 100 response in this case: https://keybase.io/_/api/1.0/user/lookup.json?usernames=s&fields=cryptocurrency_addresses.

I added this error to the `errorMessages` argument of `_requestJSON`:

```
  async fetch({ schema, options }) {
    const apiVersion = this.constructor.apiVersion
    // See https://keybase.io/docs/api/1.0/call/user/lookup.
    const url = `https://keybase.io/_/api/${apiVersion}/user/lookup.json`

    return this._requestJson({
      url,
      schema,
      options,
      errorMessages: {
        100: 'bad list value',
      }
    })
  }
}
```

But it doesn't seem to catch it:

```
[server] (node:10480) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'length' of undefined                                                                                                     
[server]     at KeybaseBTC.handle (/home/skyper/Documents/shields/services/keybase/keybase-btc.service.js:70:19)                                                                                                  
[server]     at <anonymous>
[server]     at process._tickDomainCallback (internal/process/next_tick.js:228:7)
```

The `property '[...] length' of undefined` error (line 70) comes from this piece of code:

```
 65     const data = await this.fetch({                                                                                                                                                                            
 66       schema: bitcoinAddressFoundOrNotFound,
 67       options,
 68     })
 69 
 70     if (data.them.length === 0) {
 71       return {
 72         message: 'profile not found',
 73         color: 'critical',
 74       }
 75     }
```

However, knowing that the API response doesn't contain the field `them` at all, it shouldn't match the two schemas I defined and consequently, shouldn't execute this chunk of code.

I also have a similar error with `const bitcoinAddresses = data.them[0].cryptocurrency_addresses.bitcoin`. Some profiles don't have the `cryptocurrency_addresses` field at all and it is the case of `sky`: https://keybase.io/_/api/1.0/user/lookup.json?usernames=sky&fields=cryptocurrency_addresses. This request is sent while typing `skyplabs` and similarly to the other error described earlier, my schemas let it pass.

Could you please have a look at my schemas? I have no experience with `Joi` and all of my attempts to fix these issues failed. Thank you in advance.

Reply:
> Could you please have a look at my schemas?

Sure! If you haven't already done so, can you push your latest changes? I'll take a look later today

Reply:
> Sure! If you haven't already done so, can you push your latest changes? I'll take a look later today

Thanks! I just pushed my changed (only for Keybase BTC for the moment).

> The keybase API seems to return an HTTP 100 response in this case

My bad. Even in this case, the API returns an HTTP 200 response containing an arbitrary status code 100 in the response body. The layout of my JSON viewer plug-in misled me.

Reply:
The problem is that the API response on a missing profile returns an array for `them` with the first item being a `null` (that's why the original check I proposed was `if (data.them[0] === null) {`) . If there's other scenarios where `them` is an empty array then we can update the first check accordingly to account for both scenarios

```js
if (data.them.length === 0 || !data.them[0]) {
  return {
    message: 'profile not found',
    color: 'critical',
  }
}
```

The tests are passing in my local env with the above:
```sh
  KeybaseBTC
    existing bitcoin address
      ✓
        [ GET /skyplabs.json ] (389ms)
    unknown username
      ✓
        [ GET /skyplabsssssss.json ] (409ms)


  2 passing (1s)

```

Reply:
> First, when I use the dashboard (the one on the screenshots) to generate a bitcoin badge, a request is sent for each letter of my username I type. My username being `skyplabs`, a first request is sent for `s`

I assume you're talking about the modal badge-creation window that helps users build the correct badge url? (below screenshot from my local env after making the changes mentioned above)

![image](https://user-images.githubusercontent.com/13042488/52605917-51fa9880-2e36-11e9-99f1-1ecdeb6bfc21.png)

I'm pretty sure the request is only fired after a timeout period after the last keystroke, or after focus is removed from the parameter text boxes. Either way, I wouldn't worry about that too much.

Reply:
And lastly, regarding the schemas, I think what will work best is a single schema (shown below), and remove the secondary schema and alternatives. 

```js
const bitcoinAddressSchema = Joi.object({
  them: Joi.array()
    .items(
      Joi.object({
        cryptocurrency_addresses: Joi.object({
          bitcoin: Joi.array()
            .items(
              Joi.object({
                address: Joi.string().required(),
              }).required()
            )
            .required(),
        }).required(),
      })
        .required()
        .allow(null)
    )
    .min(0)
    .max(1)
    .required(),
}).required()
```

Reply:
> I'm pretty sure the request is only fired after a timeout period after the last keystroke, or after focus is removed from the parameter text boxes. Either way, I wouldn't worry about that too much.

Actually, I didn't mean that it was a problem. I just mentioned it because, having used it during my tests, it sent requests while I was typing my username that allowed my to identify potential issues, such as a simple `s` as username which returned an error message not documented. It is a good thing in the end.

> And lastly, regarding the schemas, I think what will work best is a single schema (shown below), and remove the secondary schema and alternatives.

Indeed, with only one schema and using the `allow(null)` method, I managed to make it work. Thanks!

Reply:
Awesome! I was browsing the API docs and came across this: https://keybase.io/docs/api/1.0/responses which details the response object, including the status code of 100 which is something I can reproduce with certain username values. It's not immediately clear to me why sometimes we receive that 100 error response, but other times there's a `them` array with a single `null` entry. 

I think we could make one final tweak to the schema to catch this and provide a better error message to the user (`profile not found` instead `invalid response data`)

You should be able to utilize our `nonNegativeInteger` schema helper from `validators`:
```js
const { nonNegativeInteger } = require('../validators')
```

Then update the schema to include the status and code (and remove the `required` on the `them` array):
```js
const bitcoinAddressSchema = Joi.object({
  status: Joi.object({
    code: nonNegativeInteger.required(),
  }).required(),
  them: Joi.array()
    .items(
      Joi.object({
        cryptocurrency_addresses: Joi.object({
          bitcoin: Joi.array().items(
            Joi.object({
              address: Joi.string().required(),
            }).required()
          ),
        })
          .required()
          .allow(null),
      })
        .required()
        .allow(null)
    )
    .min(0)
    .max(1),
}).required()
```

And then finally update the first if condition to check the value of that status code (the schema `min` check on the `them` will ensure there's a value so we don't have to include the 0 length check in the code):
```js
if (data.status.code !== 0 || !data.them[0]) {
  return {
    message: 'profile not found',
    color: 'critical',
  }
}    
```

Reply:
> I was browsing the API docs and came across this: https://keybase.io/docs/api/1.0/responses which details the response object, including the status code of 100 which is something I can reproduce with certain username values.

Oh I didn't pay much attention to the number of the error code described on this page when I started my work. So, the status code 100 is well referenced in their documentation unlike what I said. These errors are pretty rare so it didn't ring a bell.

> I think we could make one final tweak to the schema to catch this and provide a better error message to the user (`profile not found` instead `invalid response data`)

Agreed. However, knowing that a status code 100 is semantically different from a status code 0 with an empty array, I would suggest returning `invalid username` instead of `profile not found` in such cases.

Reply:
Implemented.

Reply:
Awesome, thanks!